### ISSUE TITLE
Strange compilation error

### ISSUE THREAD

**Issue by hutou** [crystal-lang/crystal#8579]
Under https://play.crystal-lang.org/, the following code compiles and run fine, with either version 0.31.1 or 0.32 
```cr
tdata = [] of Array(String | Int32)
tdata << [ "abc", "def"]
tdata << [ "lmn", 3]
puts tdata
puts typeof(tdata)
```
and output is 
```
[["abc", "def"], ["lmn", 3]]
Array(Array(Int32 | String))
```
However, when running it on my PC with 'crystal play' (0.32), I get the following error :

```
Bug

You've reached a bug in the playground. Please let us know about it. 
Exception

Module validation failed: Function return type does not match operand type of return inst!
   ret i32* %1, !dbg !11
  %"Array(Int32 | String)"*
```
**but** if I run it with 'crystal run' or 'crystal build', compilation and output are Ok.

I came to this issue because of a similar exception occuring in a project I am working on and containing very similar statements.
Error message is :

```
Module validation failed: Function return type does not match operand type of return inst!
   ret i32* %1, !dbg !24
  %"Array(Float64 | Int32 | String)"* (Exception)
 Unhandled exception: Error opening file 'crystal' with mode 'r': No such file or directory (Errno)
 Failed to raise an exception: END_OF_STACK
 [0x55f4febc09f6] ???
 [0x55f4fe1f4f2b] __crystal_raise +43
 [0x55f4fe1f6725] ???
 [0x55f4fe200b8b] ???
 [0x55f4fe1fb7fc] ???
 [0x55f4fe1fac9e] ???
 [0x55f4fe56ae10] ???
 [0x55f4fe20e25f] ???
 [0x55f4fe1f8ab7] main +55
 [0x7f6244cd3153] __libc_start_main +243
 [0x55f4fe1f355e] _start +46
 [0x0] ???
```
I couldn't spot the exact origin of the problem, but noticed the error only occurs when compiling the whole program. Compilation is Ok if I compile separately the file containing these statements, and compilation is also Ok if I compile the whole program with these statements commented.

This issue did not appear under 0.31.1

My environment is :
```
$ uname -a
Linux xps15al 5.4.2-arch1-1 #1 SMP PREEMPT Thu, 05 Dec 2019 12:29:40 +0000 x86_64 GNU/Linux
$ crystal -v
Crystal 0.32.0 (2019-12-12)

LLVM: 9.0.0
Default target: x86_64-pc-linux-gnu
```

**Comment by waj** [comment]
This is the minimal code I found so far, to produce the same error without the playground:

```crystal
foo = [] of Array(String | Int32)
foo.to_s
foo << ["abc"]
```

**Comment by asterite** [comment]
Also something that shouldn't compile at all and can be compiled with `--prelude=empty`:

```crystal
class Gen(T)
end

ptr = Pointer(Gen(String | Int32)).new(0)
ptr.value = Gen(String).new
```

Eseentially we shouldn't let `Gen(U)` be passed to `Gen(T)` when `U < T` to a pointer, but it might be apply to containers in generals. Probably related to covariance and contravariance...

**Comment by HertzDevil** [comment]
Fixed by #10224.
